### PR TITLE
 distancePair class and RecommenderServlet

### PR DESCRIPTION
**Pull request by raniarusly**
created RecommenderServlet:

implemented overall class structure
implement fetching recommendation rating
implement distancePair class to represent a pair of recommendation ID and the euclidean distance to the user preference
implement function to calculate distance between rating
sort the top recommendation based on the distance
TODO:
fetch top 5 recommendation based on the calculated distance to send to the client
implement client side

### CODE REVIEW

File: src/main/java/com/google/sps/servlets/RecommenderServlet.java [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+      int distance = calculateDistance(request, recommendation);
+      distances.add(new DistancePair(id, distance));
+    }
+    return distances;
+  }
+
+  private List<Entity> getRecommendationsByCategory(String category) {
+    Query query = new Query("recommendation").setFilter(new FilterPredicate("category", FilterOperator.EQUAL, category));
+    // Get the rating for each factor
+    for(String factor : FACTORS) {
```
Comment (MostafaGabriel) [inline src/main/java/com/google/sps/servlets/RecommenderServlet.java]:
```suggestion
    for (String factor : FACTORS) {
```

File: src/main/java/com/google/sps/servlets/RecommenderServlet.java [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+      query.addProjection(new PropertyProjection(factor, Integer.class));
+    }
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+    return datastore.prepare(query).asList(FetchOptions.Builder.withDefaults());
+  }
+
+  private int calculateDistance(HttpServletRequest request, Entity recommendation) {
+    int distance = 0;
+    // for every factor considered
+    for(String factor : FACTORS) {
```
Comment (MostafaGabriel) [inline src/main/java/com/google/sps/servlets/RecommenderServlet.java]:
```suggestion
    for (String factor : FACTORS) {
```

File: src/main/java/com/google/sps/servlets/RecommenderServlet.java [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+import java.util.List;
+import javax.servlet.annotation.WebServlet;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+@WebServlet("/recommender")
+public class RecommenderServlet extends HttpServlet {
+
+  private final String[] FACTORS = {"price", "crowd"};
```
Comment (MostafaGabriel) [inline src/main/java/com/google/sps/servlets/RecommenderServlet.java]: These are stored as `price-rating` and `crowd-rating` [ref](https://github.com/googleinterns/step241-2020/blob/master/src/main/java/com/google/sps/servlets/RecommendationServlet.java#L96-L97) Please make sure to match the property name in Datastore and the HTTP request as well for the user input as you're using the same name

**Review by MostafaGabriel — COMMENTED** [review]
Thanks Rania!

File: src/main/java/com/google/sps/servlets/RecommenderServlet.java [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+import java.util.List;
+import javax.servlet.annotation.WebServlet;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+@WebServlet("/recommender")
+public class RecommenderServlet extends HttpServlet {
+
+  private final String[] FACTORS = {"price-rating", "crowd-rating"};
```
Comment (lornaarmstrong) [inline src/main/java/com/google/sps/servlets/RecommenderServlet.java]: The name "price-rating" is not correct - take a look at the datastore to check. This effects lots of the other areas of code, as when you search the datastore for this field, it finds nothing.

**Review by lornaarmstrong — CHANGES_REQUESTED** [review]

**Review by lornaarmstrong — APPROVED** [review]
LGTM!

File: src/main/java/com/google/sps/servlets/RecommenderServlet.java (hunk 1 above)
Comment (MostafaGabriel) [reply src/main/java/com/google/sps/servlets/RecommenderServlet.java]: Please fix :)

**Review by MostafaGabriel — APPROVED** [review]
LGTM! Thanks Rania :)

**Pull request merged**